### ISSUE TITLE
Integer signedness is broken

### ISSUE THREAD

**Issue by hikari-no-yume** [lynn/chibicc#9]
This old version of chibicc only has signed integers, whereas apparently the chibicc comparison opcodes (`GTH`, `LTH`) are always unsigned. We don't correct for this right now, so this means `<`, `<=`, `>` and `>=` are probably wrong when one or the other operand is signed. Considering how two's-complement works, addition, subtraction and multiplication don't have a signed/unsigned distinction, but division does, so that's also probably wrong rn.

A related thing that's not really a bug, but is annoying, is you can't use hex for two's-complement bit patterns:

```
ffff.c:3:     return 0xffff;
           ^ integer literal is too large for type int (16-bit)
```

We could change to only having unsigned integers in C, but I think it shouldn't be too hard to support both unsigned and signed with some assembly tricks.

**Comment by hikari-no-yume** [comment]
Oh wait I hadn't thought about sign-extension versus zero-extension! So currently char to short/int conversion is wrong (zero-extends, but it should sign-extend), and `>>` is wrong (doesn't preserve sign bit). These feel trickier to work around, but I'm not sure.

Supporting only unsigned integers would be easier, but it would be pretty weird…

**Comment by hikari-no-yume** [comment]
I started working on this. WIP: https://github.com/lynn/chibicc/compare/uxn...integer-signedness.

Checklist:

- Make the existing signed types behave correctly
  - [x] Sign extension
  - [x] Signed integer comparison
  - [x] Division
  - [x] Modulo
  - [x] Right shift
- [x] Put signed integer code sequences in separate functions to reduce code size
- [ ] Add `unsigned` integers (this is all type system and parsing stuff… we can take some inspiration from [modern chibicc](https://github.com/rui314/chibicc/commit/34ab83bdf49a23a47bc90354a5a4d22686d8d92a))
- [x] Write some guidance about performance/code-size caveats of signed integers

I think that's everything?

**Comment by hikari-no-yume** [comment]
Oh cool, as of 528ddd0b3b8094e8ffabb1f4e1d598f324f9d880 we're hitting a test failure related to this:

```
-1>>1
did not match expected =>
ffff
instead got
7fff
```

Update: now fixed as of 97f3a608385b01d39d3cfe61831a4961170a1997 ^^
